[MPS] Add `torch.special.bessel_[jy][01]` implementations

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149123

By copy-n-pasting functions from 
https://github.com/pytorch/pytorch/blob/f59064f2b700860a16db1930c30a4691ab663401/aten/src/ATen/native/cuda/Math.cuh#L1463

With an  ugly workaround for `bessel_y[01]` to avoid internal compiler exception on M1/M2 machines (see FB16863363 /  https://gist.github.com/malfet/e7785e4b572e7740887a83a2386ef769 )

malfet committed:
Update

[ghstack-poisoned]

Skylion007 commented on c10/metal/special_math.h:
```diff
@@ -548,5 +548,483 @@ inline T entr(T a) {
   return static_cast<T>(-INFINITY);
 }
 
```
nit add a comment about where they are copy and pasted from.

malfet commented:
Interestingly enough, `torch.special.bessel_y[01]` causes ICE on both M1 and M2 machines, but works on M4
```
>>> import torch
x=torch.rand(32, devic>>> x=torch.rand(32, device="mps")
>>> torch.special.bessel_y0(x)
tensor([-0.1166, -0.1157, -0.3485,  0.0387, -1.3865, -0.0829, -0.3282, -0.2708,
        -1.1725, -1.9554, -0.7867, -0.2320, -0.3579, -1.4972, -0.3236,  0.0721,
        -0.4343, -2.3828, -0.0327, -0.6338,  0.0818,  0.0534, -0.4901, -0.3995,
        -1.6661, -0.7198, -0.1567, -0.1664, -1.4955, -0.3342,  0.0317, -1.6034],
       device='mps:0')
>>> torch.special.bessel_y0(x>0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Failed to created pipeline state object, error: Error Domain=AGXMetalG14X Code=3 "Compiler encountered an internal error" UserInfo={NSLocalizedDescription=Compiler encountered an internal error}
```

malfet committed:
Update on "[MPS] Add `torch.special.bessel_[jy][01]` implementations"


By copy-n-pasting functions from 
https://github.com/pytorch/pytorch/blob/f59064f2b700860a16db1930c30a4691ab663401/aten/src/ATen/native/cuda/Math.cuh#L1463

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f "Lint + MPS are green"
